crypto/dilithium/p_pqdsa_asn1.c

```diff
@@ -0,0 +1,221 @@
+  if (pkey->pkey.pqdsa_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  PQDSA_KEY *key = pkey->pkey.pqdsa_key;
+  const PQDSA *pqdsa = key->pqdsa;
+
+  if (pqdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  if (key->public_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+    return 0;
+  }
+
+  if (out == NULL) {
+    *out_len = pqdsa->public_key_len;
+    return 1;
+  }
+
+  if (*out_len < key->pqdsa->public_key_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->public_key, pqdsa->public_key_len);
+  *out_len = pqdsa->public_key_len;
+  return 1;
+}
+
+static int pqdsa_pub_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // The parameters must be omitted.
+
+  // the only parameter that can be included is the OID which has length 9
+  if (CBS_len(params) != 9) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+  // set the pqdsa params on the fresh pkey
+  EVP_PKEY_pqdsa_set_params(out, OBJ_cbs2nid(params));
+  return PQDSA_KEY_set_raw_public_key(out->pkey.pqdsa_key,CBS_data(key));
+}
+
+static int pqdsa_pub_encode(CBB *out, const EVP_PKEY *pkey) {
+  PQDSA_KEY *key = pkey->pkey.pqdsa_key;
+  const PQDSA *pqdsa = key->pqdsa;
+  if (key->public_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+    return 0;
+  }
+
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // TODO: finalize this definition - OCTETSTRING to BITSTRING conversion.
+  CBB spki, algorithm, oid, key_bitstring;
+  if (!CBB_add_asn1(out, &spki, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&spki, &algorithm, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, pqdsa->oid, pqdsa->oid_len) ||
+      !CBB_add_asn1(&spki, &key_bitstring, CBS_ASN1_BITSTRING) ||
+      !CBB_add_u8(&key_bitstring, 0 /* padding */) ||
+      !CBB_add_bytes(&key_bitstring, key->public_key, pqdsa->public_key_len) ||
+      !CBB_flush(out)) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_ENCODE_ERROR);
+    return 0;
+      }
+
+  return 1;
+}
+
+static int pqdsa_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b) {
+  PQDSA_KEY *a_key = a->pkey.pqdsa_key;
+  PQDSA_KEY *b_key = b->pkey.pqdsa_key;
+
+  return OPENSSL_memcmp(a_key->public_key,
+                        b_key->public_key,
+                        a->pkey.pqdsa_key->pqdsa->public_key_len) == 0;
+}
+
+static int pqdsa_priv_decode(EVP_PKEY *out, CBS *params, CBS *key, CBS *pubkey) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 6.
+  // The parameters must be omitted.
+
+  // the only parameter that can be included is the OID which has length 9
```

Comment:
these two statement are contradicting :)

Reply:
fixed in [369080b](https://github.com/aws/aws-lc/pull/1963/commits/369080b5a595f1880b4edc75a12eb8dcbe643c02)